### PR TITLE
Show review summary in pull requests

### LINKED ISSUE #4390: Show summary of reviews
Code reviews should be shown at the sidebar of a pull request like GitHub does.

Blocked by #3748

### PR DESCRIPTION
**Pull request by kolaente**
Resolves #4390.

This pr implements a review summary on the bottom of a pr as disscussed in #4390.

It currently shows every review grouped by type and user. This means if a user rejects a pr and later approves it, both are shown in the summary (with their dates). Should I keep this behaviour or show only the newest (grouped by user)?

#### Screenshot
image

### CODE REVIEW

**Comment by kolaente** [comment]
@lafriks done.

File: models/review.go
Comment (lafriks) [inline models/review.go:269]: You can not select something that is not in group by

File: models/review.go
Comment (lafriks) [inline models/review.go]: This will also most probably fail in most db types, should be `max(review.updated_unix)`

**Review by lafriks — CHANGES_REQUESTED** [review]

File: models/review.go
Comment (kolaente) [reply models/review.go]: I've changed it so it uses the value from `select`.

File: models/review.go
Comment (kolaente) [reply models/review.go:269]: Why not? It works.

Comment (lafriks) [reply models/review.go:269]: Just tested with postgres and first I got `pq: column user.id does not exist` fixed it with escaping user table in this code:
```
Join("INNER", "`user`", "review.reviewer_id = `user`.id").
```
As I suspected now I get:
```
[E] GetReviewersByPullID: pq: column "review.type" must appear in the GROUP BY clause or be used in an aggregate function
```
I think same will be with other user table columns being selected

Comment (lunny) [reply models/review.go:269]: I think we should get the last review type since one could review multiple times?

Comment (kolaente) [reply models/review.go:269]: @lunny Thats what I do now.
@lafriks I've fixed it for postgresql, but apparantly I had to also group by `type` and then create a map afterwards to have only the newest review per user.

File: models/review.go
Comment (jonasfranz) [inline models/review.go:272]: Why exclude type `comment`?
Please use:
```sql
review.type > 0
```

File: models/review.go
Comment (jonasfranz) [inline models/review.go:273]: Why group by `review.type`? I think only the latest review of an user should be shown.

**Review by jonasfranz — CHANGES_REQUESTED** [review]

**Comment by jonasfranz** [comment]
Please add unit tests .

File: models/review.go
Comment (kolaente) [reply models/review.go:272]: > Why exclude type comment?

Because you told me to.

File: models/review.go
Comment (kolaente) [reply models/review.go:273]: Yes, but I need the query fails when using postgresql. See the comment I added in line 277.

File: models/review.go
Comment (jonasfranz) [reply models/review.go:272]: Did I? Please change it to `review.type > 0`.

Comment (kolaente) [reply models/review.go:272]: Done.

**Comment by kolaente** [comment]
@JonasFranzDEV done.

**Comment by lafriks** [comment]
@JonasFranzDEV @kolaente I asked to ignore comment as it is not relevant for approval process as if I reject or approve changes and later do some comment it should still be in state that was previously that I either rejected or approved changes

**Comment by kolaente** [comment]
@lafriks I see, I'll revert it.

**Comment by jonasfranz** [comment]
Conflicts + CI fail :drooling_face:

**Comment by lunny** [comment]
I would like use a tick icon rather than an eye icon on approval.

**Comment by kolaente** [comment]
@lunny Thats using the official function, should I change it for everythin or only in the summary? (Changing it everywhere would be a lot easier, but could also be a seperate pr)

**Comment by lunny** [comment]
@lunny I like change it for every place. OK for another PR.

**Comment by kolaente** [comment]
@lunny So this one is ready to merge?

**Comment by lunny** [comment]
@kolaente need @JonasFranzDEV  and @lafriks ' approvals.

File: models/review.go
Comment (jonasfranz) [inline models/review.go]: typo `person` instead `persion`

File: models/review_test.go
Comment (jonasfranz) [inline models/review_test.go]: Please don't create reviews since this test should only test the `GetReviwersByPullID` function. Fixtures are the right place for those reviews.

File: models/review_test.go
Comment (jonasfranz) [inline models/review_test.go]: Just add two reviews of the same users to the fixtures, since that's what you're testing here. If not, please let me know.

**Review by jonasfranz — CHANGES_REQUESTED** [review]

**Review by jonasfranz — APPROVED** [review]

**Comment by kolaente** [comment]
offtopic: why do we have `lgtm/done` if only @JonasFranzDEV has approved? :thinking:

File: models/review.go
Comment (lafriks) [inline models/review.go]: Would be better to reuse it also later if it would return array with latest statuses not map, in map this could lose actual order of reviews

Comment (lafriks) [reply models/review.go]: Yes, you can select in reverse order and than just use map[int64]bool to keep track of users already in array. And later sort.Reverse to get in correct order

Comment (kolaente) [reply models/review.go]: Done.

**Comment by kolaente** [comment]
@lafriks mind approving the pr?

**Review by lafriks — APPROVED** [review]

**Comment by lafriks** [comment]
@kolaente I don't mind ;)

**Comment by lunny** [comment]
@kolaente have the eye icon was changed to a tick icon? :)

**Pull request merged**
